### PR TITLE
`[roles-0.2.0]` Claim hat stream

### LINKED ISSUE #2186: Create UI interactions for "claiming a stream"
Each stream that has claimable tokens should give the user (the Hat Wearer of the Hat for the HatsAccount contract which is the stream recipient), the ability to make a transaction to claim their available tokens.

### PR DESCRIPTION
**Pull request by mudrila**
This PR implements basic UI and logic for withdrawing payment to the hat wearer.
It also fixes filtering of added/edited payments as previously added payment was also considered as edited.

Closes #2186 

Testing steps:

1. Create payment for any given role, where you're a wearer. Make total length of the payment short enough so that you could easily withdraw sufficient amount of tokens (1 hour for 1000 tokens should be good).
2. Vote and execute proposal for creating payment to the role
3. Go to the role details - payment info should appear there
4. Notice "Withdraw" section appearing in the payment details
5. Enter amount to withdraw (do not put maximum amount just yet) and hit "Withdraw" - this will prompt to sign transaction
6. Make sure proper amount of tokens were sent to your wallet
7. Now, hit "max" button and hit "Withdraw" and sign transaction
8. Make sure proper amount of all available for withdrawing tokens were sent to your wallet

Basic design was grabbed from https://www.figma.com/design/9hUYtvCKjOyB92vt8ukYtD/Product-Handoff?node-id=4087-18066&m=dev but not implemented 1-1, since that design should change (afaik)
Screenshot 2024-08-02 at 16 33 30

### CODE REVIEW

**Comment by tomstuart123** [comment]
Nice @mudrila - I played with this [at this link](https://deploy-preview-2201.app.dev.decentdao.org/roles?dao=sep:0x7bABf83D1836DFB0eEAADF697cAb8A7ebDc2cE5E) on the COO role - 

Testing steps:

- - Create payment for any given role, where you're a wearer. ✅
- Make total length of the payment short enough so that you could easily withdraw sufficient amount of tokens (1 hour for 1000 tokens should be good).✅
- Vote and execute proposal for creating payment to the role✅
- Go to the role details - payment info should appear there✅
- Notice "Withdraw" section appearing in the payment details✅
[- NOTE cool here was instantly I saw withdraw but no CTA. Then when token became avaiable after 1 min, it appeared. NICE] ✅
- Enter amount to withdraw (do not put maximum amount just yet) and hit "Withdraw" - this will prompt to sign transaction ✅
- Make sure proper amount of tokens were sent to your wallet ❌
- Now, hit "max" button and hit "Withdraw" and sign transaction ❌ 
- Make sure proper amount of all available for withdrawing tokens were sent to your wallet ❌

So it was all going soooo smooth. But then it seems the withdraw button isn't working right.

A few things:
--> 'MAX' doesn't seem to be giving the max tokens. It shows me one instead of the correct 4 ([video link here](https://decentdao.slack.com/archives/C06QZTUTKRC/p1722614033710269?thread_ts=1722614017.545249&cid=C06QZTUTKRC))
--> Also my wallet didn't receive the correct amount and the claim amount isn't going down. Finally, when I try and sign on metamask it also seems that no token transfer is made (see screenshot below)
Screenshot 2024-08-02 at 16 55 42

File: src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx [hunk 1]
```diff
@@ -37,18 +48,142 @@ function AccordionItemRow({ title, value, children }: AccordionItemRowProps) {
...
+            functionName: 'withdrawMax',
+            args: [bigIntStreamId, getAddress(roleHat.wearer)],
+          });
+        } else {
+          hatsAccountCalldata = encodeFunctionData({
+            abi: SablierV2LockupLinearAbi,
+            functionName: 'withdraw',
+            args: [bigIntStreamId, getAddress(roleHat.wearer), withdrawAmount.bigintValue],
+          });
+        }
```
Comment (Da-Colon) [inline src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx:145]: Does these transactions actually need to come from the Smart Account's? Might be why @tomstuart123 sees no transfer? (surprised it doesn't just revert?)

**Comment by tomstuart123** [comment]
@mudrila - cheers for tweaking the 'Max' button. This works well now

Re: the withdraw itself, sadly still doesn't work (i.e. no tokens added to my wallet and no tokens removed from claimable amount ). 

To debug - The address I'm logged in with (and receipient) is 0xC8375cC2C5fA34eba6aC386fECCaBaE1E51158c8. The Tx hash for the transaction is [here](https://sepolia.etherscan.io/tx/0x8d908340a79e2b46a8fc96364578244ff1787bfc20d6193f279b04e672393313). In etherscan it clearly is called by my address and interacts with this address (https://sepolia.etherscan.io/address/0xc12f542d81b66e02ffaf1008a680eaa3a90e30a8). 

I'll try again now with a new role, stream and address and report below. See if it works from scratch

**Comment by tomstuart123** [comment]
hey, @mudrila for some reason its now working. Not sure what is up with that current role.

but ye for the last two roles in the list [here](https://deploy-preview-2201.app.dev.decentdao.org/roles/details?dao=sep:0x7bABf83D1836DFB0eEAADF697cAb8A7ebDc2cE5E&hatId=0x000001a600010005000000000000000000000000000000000000000000000000) my address here 0xf26d75b2Fc048F1F6C793aE9F939Fa0946a0E104 has been able to claim

Feedback on gaps:
1) 'Remaining' amount isn't updating as I claim. In the below screenshot, I have claimed once from this already and yet the total amount (60) is showing with nothing removed. Note though 'withdrawable ' is working fine 
Screenshot 2024-08-05 at 16 02 29

2) A small copy error. It should be 'hang tight' not 'hand tight'
Screenshot 2024-08-05 at 16 03 42

**Comment by tomstuart123** [comment]
Hey @mudrila . These two both look good

Last thing: if you 1) click withdraw 2) then cancel it via the wallet 3) the toaster for 'hang tight' doesn't ever go away. You have to refresh page. 

Is there a way to exit if the Tx is cancelled?

File: src/components/pages/Roles/RolesDetailsDrawer.tsx [hunk 2]
```diff
@@ -43,6 +43,7 @@ interface RoleDetailsDrawerProps {
     name: string;
     wearer: string;
     description: string;
+    smartAddress: Address;
```
Comment (adamgall) [inline src/components/pages/Roles/RolesDetailsDrawer.tsx:46]: Doesn't look like this is used?

File: src/components/pages/Roles/RolesDetailsDrawerMobile.tsx [hunk 3]
```diff
@@ -15,6 +15,7 @@ interface RoleDetailsDrawerMobileProps {
     name: string;
     wearer: string;
     description: string;
+    smartAddress: Address;
```
Comment (adamgall) [inline src/components/pages/Roles/RolesDetailsDrawerMobile.tsx:18]: Doesn't look like this is used?

File: src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx [hunk 4]
```diff
@@ -37,18 +46,151 @@ function AccordionItemRow({ title, value, children }: AccordionItemRowProps) {
...
+        // @todo - Add proper copy
+        const withdrawToast = toast('Withdrawing your payment, hang tight', {
```
Comment (adamgall) [inline src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx:146]: Let's add this string, and the others below in this file, into the locales.

**Review by adamgall — CHANGES_REQUESTED** [review]

File: src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx [hunk 5]
```diff
@@ -160,17 +160,17 @@ export default function PaymentDetails({
...
           await loadAmounts();
-          toast('Payment successfully withdrawn. Check your wallet :)'); // @todo - Add proper copy
+          toast('withdrawSuccessMessage'); // @todo - Add proper copy
           setWithdrawAmount({
             value: '0',
             bigintValue: 0n,
           });
         } else {
-          toast('Transaction for withdrawing your payment reverted :('); // @todo - Add proper copy
+          toast('withdrawRevertedMessage'); // @todo - Add proper copy
```
Comment (adamgall) [inline src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx]: Think these are missing the `t()` call.

Also, let's remove these `@todo`s and just get the proper copy into this PR.

File: src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx [hunk 6]
```diff
@@ -37,18 +46,155 @@ function AccordionItemRow({ title, value, children }: AccordionItemRowProps) {
...
+          toast(t('withdrawSuccessMessage'));
+          setWithdrawAmount({
+            value: '0',
+            bigintValue: 0n,
+          });
+        } else {
+          toast(t('withdrawRevertedMessage'));
+        }
+      } catch (e) {
+        if (typeof withdrawToast !== 'undefined') {
```
Comment (adamgall) [inline src/pages/daos/[daoAddress]/roles/details/PaymentDetails.tsx:172]: Also, why not just `if (withdrawToast !== undefined) {` ?

**Review by adamgall — APPROVED** [review]

**Review by Da-Colon — APPROVED** [review]
I'll throw in that I've tested and claim does indeed work.

**Review by tomstuart123 — APPROVED** [review]
@mudrila this works now. Thanks so much for dealing with all my back and forth feedback

**Comment by xraystyle1980** [comment]
@mudrila @adamgall @tomstuart123 I can't seem to create any roles with payments. My process is
- create role, save
- create proposal, execute
- add payment stream, nothing populates in dropdown

File: src/i18n/locales/en/roles.json [hunk 7]
```diff
@@ -55,5 +55,13 @@
   "cliff": "Cliff",
   "cliffSubTitle": "How long until the assets are claimable?",
   "paymentStream": "Payment Stream",
-  "payment": "Payment"
+  "payment": "Payment",
+  "withdraw": "Withdraw",
+  "withdrawHelper": "Withdrawing will transfer the funds to the recipient's address.",
+  "remaining": "Remaining",
+  "withdrawable": "Withdrawable",
+  "withdrawPendingMessage": "Withdrawing your payment, hang tight",
```
[the remark this answers is not in the document]
Comment (nicolaus-sherrill) [reply src/i18n/locales/en/roles.json:63]: @mudrila this copy will be affected by these multi-stream UI updates, so I'm not too hung up on the specific verbiage for now. My only suggestion is changing the PendingMessage to "Your withdrawal is processing. Hang tight." Let's leave the rest as is.

In regards to the toast success/reverted/error messaging, probably best to leave out the emoticons. :-)

Success Message: "Congrats, you got paid! Your withdrawal was successful."
Reverted Message: "Your withdrawal has been reverted. Your funds are still available to claim."
Error Message: "Your withdrawal failed due to an error. Please try again."


I might be missing some technical nuance around the transactions here for copy guidance, so @tomstuart123 @xraystyle1980 please QA

**Comment by xraystyle1980** [comment]
> > @mudrila @adamgall @tomstuart123 I can't seem to create any roles with payments. My process is
> > 
> > * create role, save
> > * create proposal, execute
> > * add payment stream, nothing populates in dropdown
> 
> WDYM "nothing populates in dropdown"?

The Asset dropdown for setting up a payment is empty

**Pull request merged**